File: LeetCode/0265.Paint-House-Ii/Paint-House-Ii.java
public class Solution {
    /**
     * @param costs: n x k cost matrix
     * @return: an integer, the minimum cost to paint all houses
     */
    public int minCostII(int[][] costs) {
        // write your code here
        if(costs == null || costs.length == 0 || costs[0].length == 0) return 0;
        int m = costs.length;
        int n = costs[0].length;
        int [][]f = new int [m + 1][n];
        for(int i = 0; i < n; i++){
            f[0][i] = 0;
        }
        int a = -1;
        int b = -1;
        for(int i = 1; i <= m; i++){
            a = -1;
            b = -1;
            for(int j = 0; j < n; j++){
                if(a == -1 || f[i - 1][j] < f[i - 1][a]){
                    b = a;
                    a = j;
                }
                else{
                    if(b == -1 || f[i - 1][j] < f[i - 1][b]){
                        b = j;
                    }
                }
            }
            if(a == -1) a = 0;
            if(b == -1) b = 0;
            for(int j = 0; j < n; j++){
                if(j != a){
                    f[i][j] = f[i - 1][a] + costs[i - 1][j];
                }else{
                    f[i][j] = f[i - 1][b] + costs[i - 1][j];
                }
            }
        }
        int res = Integer.MAX_VALUE; 
        for(int i = 0; i < n; i++){
            res = Math.min(res, f[m][i]);
        }
        return res;
    }
}

File: LeetCode/0265.Paint-House-Ii/Paint-House-Ii.py
class Solution(object):
    def minCostII(self, costs):
        """
        :type costs: List[List[int]]
        :rtype: int
        """
        if not costs: return 0
        n = len(costs)
        res = [(0, float('inf')), (0, float('inf'))]
        for color, cost in enumerate(costs[0]):
            if cost < res[0][1]:
                res[1] = res[0]
                res[0] = (color, cost)
            elif cost < res[1][1]:
                res[1] = (color, cost)
        for i in range(1, n):
            cur = [(0, float('inf')), (0, float('inf'))]
            for color, cost in enumerate(costs[i]):
                if color != res[0][0]:
                    ncost = cost + res[0][1]
                else:
                    ncost = cost + res[1][1]
                if ncost < cur[0][1]:
                    cur[1] = cur[0]
                    cur[0] = (color, ncost)
                elif ncost < cur[1][1]:
                    cur[1] = (color, ncost)
            res = list(cur)
            
        return res[0][1]
                
